src/main/java/com/digitalasset/testing/junit4/Sandbox.java

```diff
@@ -6,62 +6,84 @@
 import com.daml.ledger.javaapi.data.Identifier;
 import com.daml.ledger.javaapi.data.Party;
 import com.daml.ledger.rxjava.DamlLedgerClient;
-import com.daml.daml_lf_dev.DamlLf1;
 import com.digitalasset.testing.ledger.DefaultLedgerAdapter;
 import com.digitalasset.testing.ledger.SandboxManager;
 import com.google.protobuf.InvalidProtocolBufferException;
 import io.grpc.ManagedChannel;
-import org.junit.rules.ExternalResource;
-
-import java.io.IOException;
-import java.nio.file.Files;
 import java.nio.file.Path;
 import java.time.Duration;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.function.BiConsumer;
 import java.util.function.Consumer;
-
-import static com.digitalasset.testing.utils.PackageUtils.findPackage;
-import static com.digitalasset.testing.utils.SandboxUtils.damlYamlP;
-import static com.digitalasset.testing.utils.SandboxUtils.findDamlYaml;
+import org.junit.rules.ExternalResource;
 
 public class Sandbox {
-  private static Duration DEFAULT_WAIT_TIMEOUT = Duration.ofSeconds(30);
-  private static String[] DEFAULT_PARTIES = new String[] {};
+  private static final Duration DEFAULT_WAIT_TIMEOUT = Duration.ofSeconds(30);
+  private static final String[] DEFAULT_PARTIES = new String[] {};
   private final SandboxManager sandboxManager;
 
   public static SandboxBuilder builder() {
     return new SandboxBuilder();
   }
 
+  private Sandbox(
+      Path damlRoot,
+      Optional<String> testModule,
+      Optional<String> testStartScript,
+      Duration waitTimeout,
+      String[] parties,
+      Path darPath,
+      BiConsumer<DamlLedgerClient, ManagedChannel> setupApplication,
+      boolean useWallclockTime,
+      boolean useReset,
+      Optional<String> ledgerId,
+      Optional<LogLevel> logLevel) {
+    this.sandboxManager =
+        new SandboxManager(
+            damlRoot,
+            testModule,
+            testStartScript,
+            waitTimeout,
+            parties,
+            darPath,
+            setupApplication,
+            useWallclockTime,
+            ledgerId,
+            logLevel);
+    this.useReset = useReset;
+  }
+
+  private final boolean useReset;
+
   public static class SandboxBuilder {
     private Optional<String> testModule = Optional.empty();
     private Optional<String> testStartScript = Optional.empty();
     private Duration waitTimeout = DEFAULT_WAIT_TIMEOUT;
     private String[] parties = DEFAULT_PARTIES;
-    private Path projectRoot;
+    private Path damlRoot;
     private Path darPath;
     private boolean useWallclockTime = false;
     private boolean useReset = false;
-    private BiConsumer<DamlLedgerClient, ManagedChannel> setupApplication;
+    private BiConsumer<DamlLedgerClient, ManagedChannel> setupApplication = (t, u) -> {};
     private Optional<String> ledgerId = Optional.empty();
     private Optional<LogLevel> logLevel = Optional.empty();
 
     public SandboxBuilder dar(Path darPath) {
+      Objects.requireNonNull(darPath);
```

Comment:
I would put these checks in the build method. Since this will be used in a fluent API manner, the exception will come from the same call site, however, we want to throw error for both setting it to null and not setting it at all.

Reply:
This also applies for all the other checks below.

Reply:
I wanted to resolve some inconsistency and I was going back and forth between those two approaches myself as well. I thought it is clearer to throw error as early as possible. But I guess the `build` is not too far either.